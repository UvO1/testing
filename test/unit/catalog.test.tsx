import React from 'react';
import { initStore, productDetailsLoaded, productDetailsLoad, productsLoad, productsLoaded } from '../../src/client/store';
import { Application } from '../../src/client/Application';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { ExampleApi, CartApi } from '../../src/client/api';
import { render,screen,act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Catalog } from '../../src/client/pages/Catalog';
import { mock_cart, mock_first_product, mock_second_product } from './mock';
import renderer from "react-test-renderer";


describe('Catalog', () => {
    const basename = '/hw/store';
    const api = new ExampleApi(basename);
    const cart = new CartApi();

    const store = initStore(api, cart);
    jest.mock('axios');
    it('для каждого товара в каталоге отображается название, цена и ссылка на страницу с подробной информацией о товаре', async() => {
        render(  
            <MemoryRouter initialEntries={['/catalog']}>
                <Provider store={store}>
                    <Catalog/>
                </Provider>
            </MemoryRouter>
        );
        await act(() => store.dispatch(productsLoaded([mock_first_product, mock_second_product] )));
        const first_product = screen.getAllByTestId('1')[0];
        const second_product = screen.getAllByTestId('2')[0];
        expect(first_product.getElementsByClassName('ProductItem-Name')[0].innerHTML).toBe(mock_first_product.name);
        expect(first_product.getElementsByClassName('ProductItem-Price')[0].innerHTML).toBe('$' + mock_first_product.price);
        expect(first_product.getElementsByClassName('ProductItem-DetailsLink')[0].getAttribute("href")).toBe('/catalog/1');
        expect(second_product.getElementsByClassName('ProductItem-Name')[0].innerHTML).toBe(mock_second_product.name);
        expect(second_product.getElementsByClassName('ProductItem-Price')[0].innerHTML).toBe('$' + mock_second_product.price);
        expect(second_product.getElementsByClassName('ProductItem-DetailsLink')[0].getAttribute("href")).toBe('/catalog/2');
    });

    it('на странице с подробной информацией отображаются: название товара, его описание, цена, цвет, материал и кнопка "добавить в корзину"', async() => {
        const domTree = renderer.create(  
            <MemoryRouter initialEntries={['/catalog/1']}>
                <Provider store={store}>
                    <Application/>
                </Provider>
            </MemoryRouter>
        );
        await act(() => store.dispatch(productDetailsLoad(1)));
        await act(() => store.dispatch(productDetailsLoaded(mock_first_product)));
        expect(domTree).toMatchSnapshot();
    });

    it('если товар уже добавлен в корзину, на странице товара должно отображаться сообщение об этом', () => {
        cart.setState(mock_cart); //товар под id=1 добавлен в корзину
        const store = initStore(api, cart);

        //будем наблюдать за страницей товара под id=1 
        render(  
            <MemoryRouter initialEntries={['/catalog/1']}> 
                <Provider store={store}>
                    <Application/>
                </Provider>
            </MemoryRouter>
        );

        act(() => store.dispatch(productDetailsLoad(1)));
        act(() => store.dispatch(productDetailsLoaded(mock_first_product))); //имитируем подгрузку информации о товаре на страницу
        expect(screen.getByText('Item in cart')).not.toBeUndefined(); //находясь на странице товара под id=1 есть информация, что он добавлен в корзину
    });
    

});

