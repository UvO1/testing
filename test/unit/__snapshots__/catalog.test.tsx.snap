// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog на странице с подробной информацией отображаются: название товара, его описание, цена, цвет, материал и кнопка "добавить в корзину" 1`] = `
<div
  className="Application"
>
  <nav
    className="navbar navbar-expand-sm navbar-light bg-light"
  >
    <div
      className="container"
    >
      <a
        className="Application-Brand navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Kogtetochka store
      </a>
      <button
        aria-label="Toggle navigation"
        className="Application-Toggler navbar-toggler"
        onClick={[Function]}
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="Application-Menu collapse navbar-collapse"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current="page"
            className="active nav-link"
            href="/catalog"
            onClick={[Function]}
          >
            Catalog
          </a>
          <a
            className="nav-link"
            href="/delivery"
            onClick={[Function]}
          >
            Delivery
          </a>
          <a
            className="nav-link"
            href="/contacts"
            onClick={[Function]}
          >
            Contacts
          </a>
          <a
            className="nav-link"
            href="/cart"
            onClick={[Function]}
          >
            Cart
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container pt-4"
  >
    <div
      className="Product"
    >
      <div
        className="ProductDetails row"
      >
        <div
          className="col-12 col-sm-5 col-lg-4"
        >
          <img
            className="Image"
            src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
          />
        </div>
        <div
          className="col-12 col-sm-7 col-lg-6"
        >
          <h1
            className="ProductDetails-Name"
          >
            Practical kogtetochka
          </h1>
          <p
            className="ProductDetails-Description"
          >
            Really Unbranded kogtetochka for Snowshoe
          </p>
          <p
            className="ProductDetails-Price fs-3"
          >
            $
            685
          </p>
          <p>
            <button
              className="ProductDetails-AddToCart btn btn-primary btn-lg"
              onClick={[Function]}
            >
              Add to Cart
            </button>
          </p>
          <dl>
            <dt>
              Color
            </dt>
            <dd
              className="ProductDetails-Color text-capitalize"
            >
              Ivory
            </dd>
            <dt>
              Material
            </dt>
            <dd
              className="ProductDetails-Material text-capitalize"
            >
              Wooden
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
</div>
`;
